htap.go

```diff
@@ -22,5 +25,41 @@ func (qcc *queryContextCache) init() *queryContextCache {
 func (qcc *queryContextCache) add(qces ...queryContextEntry) {
 	qcc.mutex.Lock()
 	defer qcc.mutex.Unlock()
-	qcc.entries = append(qcc.entries, qces...)
+	var newQcesBasedOnPriorities []queryContextEntry
+	for _, newQce := range qces {
+		logger.Debugf("adding query context: %v", newQce)
+		newQceProcessed := false
+		for existingQceIdx, existingQce := range qcc.entries {
+			if newQce.ID == existingQce.ID {
+				newQceProcessed = true
+				if newQce.Timestamp > existingQce.Timestamp {
+					if newQce.Priority == existingQce.Priority {
+						qcc.entries[existingQceIdx].Timestamp = newQce.Timestamp
+						qcc.entries[existingQceIdx].Context = newQce.Context
+					} else {
+						qcc.entries[existingQceIdx] = newQce
+					}
+				} else if newQce.Timestamp == existingQce.Timestamp {
+					if newQce.Priority != existingQce.Priority {
+						qcc.entries[existingQceIdx].Priority = newQce.Priority
+						qcc.entries[existingQceIdx].Context = newQce.Context
+					}
+				}
+			} else {
+				if newQce.Priority == existingQce.Priority {
+					qcc.entries[existingQceIdx].ID = newQce.ID
```

Comment:
cannot we just override the whole item in the list?

Reply:
Let's try this out :)